src/emqx_protocol.erl

```diff
@@ -589,13 +589,16 @@ deliver({disconnect, _ReasonCode}, PState) ->
 -spec(send(emqx_mqtt_types:packet(), state()) -> {ok, state()} | {error, term()}).
 send(Packet = ?PACKET(Type), PState = #pstate{proto_ver = Ver, sendfun = SendFun}) ->
     trace(send, Packet, PState),
-    case SendFun(emqx_frame:serialize(Packet, #{version => Ver})) of
+    case SendFun(Packet, #{version => Ver}) of
         ok ->
             emqx_metrics:sent(Packet),
             {ok, inc_stats(send, Type, PState)};
         {binary, _Data} ->
             emqx_metrics:sent(Packet),
             {ok, inc_stats(send, Type, PState)};
+        {datagram, _Peer, _Data} ->
```

Comment:
This clause and the `{binary, _Data}` clause should be merged into `ok` (make `SendFun` return `ok` instead).

Reply:
The websocket SendFun returns `{binary, Data}` only because the last line is `Pid ! {binary, Data}`.
It's silly to return the send result to the caller then discarded here anyway.

Reply:
Yes, can't agree more.